File: python/random/166.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        sign = "-" if numerator * denominator < 0 else ""

        numerator = abs(numerator)
        denominator = abs(denominator)

        res = f"{sign}{numerator // denominator}"
        r = numerator % denominator

        if r == 0:
            return res
        res += "."

        unique = {}
        seq = []
        idx = -1
        while r > 0:
            if r in unique:
                res += ''.join(str(n) for n in seq[:unique[r] + 1])
                res += f"({''.join(str(n) for n in seq[unique[r] + 1:])})"
                return res

            unique[r] = idx
            while r < denominator:
                r *= 10
                idx += 1
                unique[r] = idx
                seq.append(0)

            seq[idx] = r // denominator
            r %= denominator

        res += ''.join(str(n) for n in seq)
        return res


def test():
    print(Solution().fractionToDecimal(2, 7))
    print(Solution().fractionToDecimal(1, 9))
    print(Solution().fractionToDecimal(1, 2))
    print(Solution().fractionToDecimal(2, 1))
    print(Solution().fractionToDecimal(4, 333))
    print(Solution().fractionToDecimal(1 + 50 * 900, 900))
    print(Solution().fractionToDecimal(719, 990))
    print(Solution().fractionToDecimal(1, -8))


if __name__ == '__main__':
    test()

File: go/problems/166.go
package problems

import (
	"fmt"
	"strconv"
	"strings"
)

func fractionToDecimal(numerator int, denominator int) string {
	res := ""

	abs := func(a int) int {
		if a < 0 {
			return -a
		}
		return a
	}

	if numerator*denominator < 0 {
		res += "-"
	}
	numerator = abs(numerator)
	denominator = abs(denominator)

	res += strconv.Itoa(numerator / denominator)
	rem := numerator % denominator
	if rem == 0 {
		return res
	}

	rems := make(map[int]int)
	seq := []int{}
	idx := -1

	for rem > 0 {
		if _, ok := rems[rem]; ok {
			var prx, tail strings.Builder
			for _, v := range seq[:rems[rem]+1] {
				prx.WriteString(strconv.Itoa(v))
			}

			for _, v := range seq[rems[rem]+1:] {
				tail.WriteString(strconv.Itoa(v))
			}

			return fmt.Sprintf("%s.%s(%s)", res, prx.String(), tail.String())
		}

		rems[rem] = idx
		for rem < denominator {
			rem *= 10
			seq = append(seq, 0)
			idx++
		}

		seq[idx] = rem / denominator
		rem = rem % denominator
	}

	var t strings.Builder
	for _, v := range seq {
		t.WriteString(strconv.Itoa(v))
	}

	return fmt.Sprintf("%s.%s", res, t.String())
}
